Throw exception when reading from AsyncBody after the channel closed (#2399)

lackhoa opened the pull request:
Currently, when we call `Response.body.asStream`, we don't check to see if the corresponding response channel is open or not. If the channel has been closed and the user attempts to read the response body, it's possible for the body stream to hang (i.e never finish or throw an error), leading to the code hanging as illustrated in #2399. This behavior is undesirable since the user isn't aware of the channel closing (because we take care of that behind the scene), leading to the issue being hard to debug.

This commit adds a check to `AsyncBodyReader.connect` so that we will throw an exception when the user attempts to read an unfinished response body from a closed channel.

PR note: I'm new to the code base, so let me know if throwing an exception is bad style (I'm doing it because there's already an exception in that same function :)

/claim #2399

lackhoa committed:
Throw exception when reading from AsyncBody after the channel closed (#2399)

Currently, when we call `Response.body.asStream`, we don't check to see if the
corresponding response channel is open or not. If the channel has been closed
and the user attempts to read the response body, it's possible for the body stream
to hang (i.e never finish or throw an error), leading to the code hanging as
illustrated in #2399. This behavior is undesirable since the user isn't aware of the
channel closing (because we take care of that behind the scene), leading to the
issue being hard to debug.

This commit adds a check to `AsyncBodyReader.connect` so that we will throw
an exception when the user attempts to read an unfinished response body from
a closed channel.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/zio/zio-http?pullRequest=2454) <br/>All committers have signed the CLA.

myazinn commented on zio-http/src/main/scala/zio/http/netty/AsyncBodyReader.scala:
```diff
@@ -38,11 +38,22 @@ abstract class AsyncBodyReader(implicit trace: Trace) extends SimpleChannelInbou
           }
-          ctx.read()
+
+          if (ctx.channel.isOpen || readingDone) {
```
It looks like a feature, but just want to highlight that it'll lead to cases when we can read from a closed channel. It is possible that reading will be done before we try to connect to a body, thus leading to confusing flaky behaviour when request sometimes succeeds, sometimes doesn't.
E.g.
```scala
package example

import zio._
import zio.http._

object SimpleClient extends ZIOAppDefault {
  def run: RIO[Scope, Any]               = program.provideSome[Scope](Client.default)
  def program: RIO[Scope & Client, Unit] = for {
    client <- ZIO.service[Client]
    request = client.request(Request.get("https://jsonplaceholder.typicode.com/todos/1"))
    res <- ZIO.scoped(request)
    _   <- res.body.asString.debug
  } yield ()
}
```
It should fail IMO, but on my machine it succeeds most of the time.
BTW latest version of bio-http sometimes fails with mysterious 
```
timestamp=2023-09-23T14:16:20.575404Z level=ERROR thread=#zio-fiber-1 message="" cause="Exception in thread "zio-fiber-19,4" io.netty.channel.AbstractChannel$AnnotatedNoRouteToHostException: No route to host: jsonplaceholder.typicode.com/[2606:4700:e4:0:0:0:ac40:a206]:443
```
Which is not related to your changes (looks like another bug, lol)

lackhoa replied:
I'm aware of the flakiness you suggested. But the check I added there is not a feature so much as a requirement. Like, is it *required* to keep the channel open while we read from the buffer? (ie Is it the case that in the future, our code will never close the channel before `AsyncBodyReader.connect` is called?)

If that's the case, I'm all for failing whenever the channel is closed. If not, then the check is required.

And I agree that your example should fail, but a fix for that can be proposed in another PR.
(For future context, the mentioned flakiness refers to *broken* code, not *correct* ones)

jdegoes commented:
@lackhoa Thanks for your work on this! Please add a regression test based on the original issue to ensure the issue is succcessfully fixed and won't regress in the future. 🙏

lackhoa commented:
> @lackhoa Thanks for your work on this! Please add a regression test based on the original issue to ensure the issue is succcessfully fixed and won't regress in the future. 🙏

It hath been done!
Edit: Ok I spoke too soon, CI was angry at me?

lackhoa committed:
fix: AsyncBodyReader emit failure when channel closes

This is a bug which happened to me in CI: when the connect to the server closes,
we must fail the response body stream so that the client doesn't hang forever
trying to read from the stream.

lackhoa commented:
@jdegoes Ok test issue is fixed now, turned out there was another bug lurking in the code.

adamgfraser reviewed (changes requested):
I don't think this works. We are checking whether the channel is open immediately before reading from it, but there is a concurrency problem here. The channel may be open when we check it but then closed before we read from it.

lackhoa commented:
@adamgfraser I’ve added a channelInactive to AsyncBodyReader as well. I don’t remember exactly, but that should cover all the cases.
